incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -15,6 +15,201 @@ import (
 	api "sigs.k8s.io/multi-tenancy/incubator/hnc/api/v1alpha2"
 )
 
+var _ = Describe("Exception", func() {
+	ctx := context.Background()
+
+	BeforeEach(func() {
+		// We want to ensure we're working with a clean slate, in case a previous tests objects still exist
+		cleanupObjects(ctx)
+	})
+
+	AfterEach(func() {
+		// Change current config back to the default value.
+		resetHNCConfigToDefault(ctx)
+		cleanupObjects(ctx)
+	})
+
+	Context("Add exception annotations ", func() {
+		const (
+			p  = "parent"
+			c1 = "child1"
+			c2 = "child2"
+			c3 = "child3"
+		)
+		tests := []struct {
+			name         string
+			selector     string
+			treeSelector string
+			want         []string
+			notWant      []string
+		}{
+			{name: "simply using select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "using multiple select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix + ", !" + c2 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "simply using treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1,
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "using multiple treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1 + ", !" + c2,
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "using both select and treeSelect negatively",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "!" + c2,
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "using both select and treeSelect positively",
+				selector:     c1 + api.LabelTreeDepthSuffix,
+				treeSelector: c2,
+				want:         []string{},
+				notWant:      []string{c1, c2, c3},
+			},
+		}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should handle exceptions: "+tc.name, func() {
+				// Set up namespaces
+				fooName := createNS(ctx, "foo")
+				barName := createNS(ctx, "bar")
+				bazName := createNS(ctx, "baz")
+				quxName := createNS(ctx, "qux")
```

Comment:
Suggestion: line these up with the constants used to define the test, i.e. `child1`, etc.

Reply:
Another suggestion:
```
names := map[string]string{
  p:  createNS(ctx, p),
  c1: createNS(ctx, c1),
  c2: createNS(ctx, c2),
  c3: createNS(ctx, c3),
}

setParent(ctx, names[c1], names[p])
setParent(ctx, names[c2], names[p])
setParent(ctx, names[c2], names[p])
tc.selector = replaceStrings(tc.selector, names)
tc.treeSelector = replaceStrings(tc.treeSelector, names)
```

Reply:
done